cmd_watch_dir.go

```diff
@@ -0,0 +1,68 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+)
+
+// CmdWatchDir is `direnv watch-dir SHELL PATH`
+var CmdWatchDir = &Cmd{
+	Name:    "watch-dir",
+	Desc:    "Recursively adds a directory to the list that direnv watches for changes",
+	Args:    []string{"[SHELL]", "[DIR]"},
+	Private: true,
+	Action:  actionSimple(watchDirCommand),
+}
+
+func watchDirCommand(env Env, args []string) (err error) {
+	var shellName string
+	var dir string
+
+	if len(args) < 2 {
+		return fmt.Errorf("a directory is required to add to the list of watches")
+	}
+	if len(args) > 2 {
+		shellName = args[1]
+		dir = args[2]
+	} else {
+		shellName = "bash"
+		dir = args[1]
+	}
+
+	shell := DetectShell(shellName)
+
+	if shell == nil {
+		return fmt.Errorf("unknown target shell '%s'", shellName)
+	}
+
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		return fmt.Errorf("dir '%s' does not exist", dir)
+	}
+
+	watches := NewFileTimes()
+	watchString, ok := env[DIRENV_WATCHES]
+	if ok {
+		err = watches.Unmarshal(watchString)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
```

Comment:
```suggestion
if info.Isdir() {
  return nil
}
```
We should only be watching files as folders don't have meaningful mtimes.

Reply:
Is it not worth keeping directories as part of this? meaning that we'll reload `direnv` whenever we add/remove a file from the directory (or a subdirectory). I'm not sure how mtimes work so I'm probably wrong here :)

Reply:
I have added this suggestion for now though

Reply:
Actually, I don't know.

I assumed that the directory didn't have a meaningful mtime. But now that I am testing it locally, it looks like it *does* update whenever a file gets created. So I might be wrong. It looks like it works on Linux, but I don't know if it's portable across POSIX.

Reply:
Not only it does but it also has to: https://serverfault.com/a/388100 (and yep looks like POSIX is what IEEE 1003 is).